### DISCUSSION TITLE
Is there a way to get the physical operand size from a ZydisDecodedOperand?

### DISCUSSION THREAD

**Discussion by jonomango** [zyantific/zydis#356] [Q&A] [answered]
Hi,

Sorry for asking so many questions, but I was wondering if there was a way to get the physical operand size (specifically for imm/disp operands) from a decoded operand. I've found myself often working with a `ZydisDecodedOperand` structure that I need the size of, and it is a bit annoying to have to get this information from the `ZydisDecodedInstructionRaw` structure. This isn't *that* big of a problem until you start working with formatter hooks (like `ZYDIS_FORMATTER_FUNC_FORMAT_OPERAND_MEM` and `ZYDIS_FORMATTER_FUNC_FORMAT_OPERAND_IMM`) and you don't know which field to access in the raw decoded instruction field (`imm[0]` vs `imm[1]`, for example).

Any help would be appreciated, and again, sorry for making so many discussion posts :sweat_smile:

**Comment by flobernd** [comment, marked as answer]
Hey,

@athre0z is currently working on refactoring the whole `.raw` stuff. In fact we are trying to remove most of the stuff from `ZydisDecodedInstruction` and provide a different API to query it. The offsets and sizes to the individual segments (prefixes, imm, disp, ...) will still be available tho and it would be a step in the "wrong direction" to add the raw offset/size to `ZydisDecodedOperand`.

I'm curious: Is there a particular reason you have to access the raw values in the formatter? Usually the formatter aims to hide any encoding specific details.

To work around your specific problem you could introduce a simple `int` in your context which you set to `0` when a new instruction is being formatted and increase by `1` everytime an immediate operand is formatted. I admit, it's not a perfect solution as you have to e.g. reverse the order for ATT output, but it's basically the same way the decoder does it internally in e.g. `ZydisDecoderDecodeOperand`.

**Reply by jonomango** [reply]
The reason I need the physical size is so that I can calculate the unsigned value for the imm/disp operand. It's very ugly, but basically I do:

```cpp
auto const mask = (1ull << physical_operand_size) - 1;
auto const unsigned_value = context->operand->imm.value.u & mask;
```

This is needed because the operand value is sign-extended to 64-bits, which completely changes the unsigned value for negative numbers. For example:

```cpp
std::int32_t value_32 = 0xFFFFFFF5;
std::int16_t value_16 = 0xFFF5;
```

These have the same value (-11) when treated as signed numbers, but are different when treated as unsigned. This matters a lot for me since I'm re-encoding memory references to use 32-bit unsigned symbol IDs instead. A simple `jmp short +0xF5` would actually be jumping to the location at symbol `0xF5`, if that makes sense. These symbols have names, so *ideally* I would be able to just format with `symbol_table[context->operand->imm.value.u]->name`, but I can't because of the signed/unsigned issue.

**Reply by flobernd** [reply]
Oh, I see. But can't you just use the `ZYDIS_FORMATTER_FUNC_PRINT_ADDRESS_ABS` hook instead like demonstrated in this example:
https://github.com/zyantific/zydis/blob/master/examples/Formatter01.c

The formatter has been designed with symbol printing in mind. As well there is the `ZydisCalcAbsoluteAddress()` utility function which should be used to convert relative adresses to absolute ones (e.g. in order to look up a symbol).

**Reply by jonomango** [reply]
I *think* I could, but remember, I'm encoding the symbol ID as the relative offset, so `ZydisCalcAbsoluteAddress` wouldn't be able to calculate the correct symbol ID. I would have to pass 0 as the `runtime_address` (which is impossible), then also subtract the instruction length from the return value in order to get the "correct" value, and even then, I don't think I would end up with the value I'm looking for because of the signed/unsigned B.S. 😅

**Reply by flobernd** [reply]
> I'm encoding the symbol ID as the relative offset

But isn't this information completely ambiguous without the `runtime_address` anyways? 😄 I probably don't completely understand what you are trying to achieve.

**Reply by jonomango** [reply]
This might explain a bit more...

![image](https://user-images.githubusercontent.com/26609800/176615228-34dbe237-50dd-4f41-866a-6a5b8a78a9fc.png)

For example, the `jnz loc_1022` instruction is actually a `jnz +0x4` instruction, where `0x4` is the symbol ID of `loc_1022`. In the formatter hook, I'm replacing `0x4` with `symbol_table[0x4]->name` so that it prints in a nice way.

**Reply by flobernd** [reply]
What confuses me is what you do in this case:

```
+000 jnz +04 # loc_1022
+002 jmp +02 # loc_1022
```

Both branches have the same destination (`loc_1022`) but different offsets. Even worse like this:

```
+000 jnz +04 # loc_1022
+0D8 jnz +04 # loc_8888
```

where both branches have the same relative offsets but different destinations.

**Reply by jonomango** [reply]
Ah, I think I see where the confusion is coming from. These aren't actually instructions that are meant to be *executed*, the way they currently are. This is just an intermediate representation for them that is *very* similar to x86. In this IR, there are no relative instructions *at all*. Every relative instruction has been modified to use a symbol ID, which is absolute. I'm just using the term "relative offset" to refer to what is now known as "symbol ID."

```
+000 jnz +04 # loc_1022
+002 jmp +02 # loc_1022
```

These do not both point to `loc_1022`, since they have different symbol IDs (remember, these are absolute, not relative). The first instruction points to `loc_1022` and the second points to `entrypoint`.

```
+000 jnz +04 # loc_1022
+0D8 jnz +04 # loc_8888
```

These both point to `loc_1022` (symbol ID 04).

Again, symbol IDs are absolute, but they're encoded in the same position that normal x86 relative operands would be. When the time comes to produce normal x86 instructions, I just look into the symbol table for the real address (if it has one).

**Reply by flobernd** [reply]
> This is just an intermediate representation for them that is very similar to x86

Alright, that was not clear to me in the beginning. 

In this case I don't have a better solution than manually keeping track of the immediate id like I suggested in my first answer. It's similar to what we do internally e.g. here:
https://github.com/zyantific/zydis/blob/master/src/Decoder.c#L1861

**Reply by jonomango** [reply]
> In this case I don't have a better solution than manually keeping track of the immediate id like I suggested in my first answer. It's similar to what we do internally e.g. here: https://github.com/zyantific/zydis/blob/master/src/Decoder.c#L1861

I see. In my situation, I'm only dealing with relative instructions, so I don't actually have to worry about which `imm` to use. From my understanding (and a decent amount of testing), there can be at most a single relative operand in an instruction and it is always the first `imm` operand. This means that I don't have to worry about `imm[1]` ever being relative.

Anyways, thanks for the help, I really appreciate it! The Zydis community has been nothing less than fantastic. ❤️

**Reply by athre0z** [reply]
> The reason I need the physical size is so that I can calculate the unsigned value for the imm/disp operand.

That's also something that we're currently working on making easier, by the way! :)

See discussion here: https://github.com/zyantific/zydis/pull/336#issuecomment-1140337971

That being said, it most likely won't make it for the v4.0 release, so implementing the suggested workaround is not a waste of time. If only days had more hours .. :p

**Reply by jonomango** [reply]
> That's also something that we're currently working on making easier, by the way! :)
> 
> See discussion here: [#336 (comment)](https://github.com/zyantific/zydis/pull/336#issuecomment-1140337971)

I'm glad to see that it is being worked on! I'm looking forward to the new updates.
